### PR TITLE
feat: support for building packages with redhat ubi

### PR DESCRIPTION
**Pull request by Revolyssup**
Related issue: https://github.com/apache/apisix/issues/9454

### CODE REVIEW

**Comment by tao12345666333** [comment]
Please sign CLA first.

File: .github/workflows/package-apisix-base-rpm-ubi.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+      - name: run ubi8 docker and mapping apisix-base rpm into container
+        run: |
+          docker run -itd -v /home/runner/work/apisix-build-tools/apisix-build-tools/output:/output --name ubiInstance --net="host" registry.access.redhat.com/ubi8/ubi:latest /bin/bash
+
+      - name: install dependencies in container
+        run: |
+          docker exec ubiInstance bash -c "dnf install -y centos-release-stream"
+          docker exec ubiInstance bash -c "dnf swap -y centos-{linux,stream}-repos"
+          docker exec ubiInstance bash -c "dnf distro-sync -y"
```
Comment (tao12345666333) [inline .github/workflows/package-apisix-base-rpm-ubi.yml:42]: why use centos repo here?

File: .github/workflows/package-apisix-base-rpm-ubi.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: install dependencies
+        run: |
+          sudo apt-get install -y make ruby ruby-dev rubygems build-essential
+
+      - name: build apisix-base rpm
+        run: |
+          make package type=rpm app=apisix-base version=${BUILD_APISIX_BASE_VERSION} image_base=registry.access.redhat.com/ubi8/ubi image_tag=latest
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply .github/workflows/package-apisix-base-rpm-ubi.yml]: Done. I changed it to `8.8`

File: .github/workflows/package-apisix-dashboard-rpm-ubi.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+      - name: Run ubi8 docker and mapping rpm into container
+        run: |
+          docker run -itd -v $PWD/output:/apisix-dashboard --name ubiInstance --net="host" registry.access.redhat.com/ubi8/ubi:8.8 /bin/bash
+
+      - name: Install rpm package
+        run: |
+          docker exec ubiInstance bash -c "dnf install -y centos-release-stream"
+          docker exec ubiInstance bash -c "dnf swap -y centos-{linux,stream}-repos"
+          docker exec ubiInstance bash -c "dnf distro-sync -y"
```
Comment (tao12345666333) [inline .github/workflows/package-apisix-dashboard-rpm-ubi.yml]: ditto

File: .github/workflows/package-apisix-base-rpm-ubi.yml (hunk 1 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-base-rpm-ubi.yml]: removed these

File: .github/workflows/package-apisix-dashboard-rpm-ubi.yml [hunk 4]
```diff
@@ -0,0 +1,75 @@
+name: package apisix-dashabord rpm for ubi
```
Comment (tao12345666333) [inline .github/workflows/package-apisix-dashboard-rpm-ubi.yml:1]: We don't need to build an ubi package for the dashboard. 
This file can be deleted.

File: .github/workflows/package-apisix-rpm-ubi.yml [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+      - name: install dependencies
+        run: |
+          sudo apt-get install -y make ruby ruby-dev rubygems build-essential
+          sudo gem install --no-document fpm
+          sudo apt-get install -y rpm
+
+      - name: packaging APISIX(-remote) with remote code
+        run: |
+          make package type=rpm app=apisix version=master checkout=master image_base=registry.access.redhat.com/ubi8/ubi image_tag=latest artifact=apisix-remote
```
Comment (tao12345666333) [inline .github/workflows/package-apisix-rpm-ubi.yml:42]: Do we also need to change the image tag?

File: .github/workflows/package-apisix-rpm-ubi.yml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        run: |
+          git clone https://github.com/apache/apisix.git
+          make package type=rpm app=apisix version=master checkout=master image_base=registry.access.redhat.com/ubi8/ubi image_tag=8.8 local_code_path=./apisix artifact=apisix-local
+
+      - name: run ubi8 docker and mapping rpm into container
+        run: |
+          docker run -itd -v $PWD/output:/output --name ubiInstance --net="host" registry.access.redhat.com/ubi8/ubi:8.8 /bin/bash
+          docker exec ubiInstance bash -c "dnf install -y centos-release-stream"
+          docker exec ubiInstance bash -c "dnf swap -y centos-{linux,stream}-repos"
+          docker exec ubiInstance bash -c "dnf distro-sync -y"
```
Comment (tao12345666333) [inline .github/workflows/package-apisix-rpm-ubi.yml]: ditto. centos repo.

**Comment by tao12345666333** [comment]
Due to the fact that both CentOS 8 and UBI 8 have RPM package names of xxx.el8.xxx, you need to confirm whether they will conflict.

File: .github/workflows/package-apisix-dashboard-rpm-ubi.yml (hunk 4 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-dashboard-rpm-ubi.yml]: Done deleted 👍

File: .github/workflows/package-apisix-rpm-ubi.yml (hunk 5 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-rpm-ubi.yml]: Yep. I missed it here. Changed it now.

File: .github/workflows/package-apisix-rpm-ubi.yml (hunk 6 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-rpm-ubi.yml]: Already removed 👍

**Comment by Revolyssup** [comment]
> Due to the fact that both CentOS 8 and UBI 8 have RPM package names of xxx.el8.xxx, you need to confirm whether they will conflict.

By conflict, do you mean whether running `make package` for centos and ubi will create the same rpm file name in /output directory ?

**Comment by tao12345666333** [comment]
Yes. This is what you need to pay attention to.

File: utils/install-common.sh [hunk 7]
```diff
@@ -52,8 +58,8 @@ install_openresty_deb() {
 
 install_openresty_rpm() {
     # install openresty and openssl111
-    yum-config-manager --add-repo https://openresty.org/package/centos/openresty.repo
-    yum install -y openresty openresty-openssl111-devel pcre pcre-devel openldap-devel
+        yum-config-manager --add-repo https://openresty.org/package/centos/openresty.repo
```
Comment (Sn0rt) [inline utils/install-common.sh:61]: why modify this ?

**Comment by Revolyssup** [comment]
> Details

Here[1] the default is set as `el7`. So currently the package name in output is `*.el7*.rpm`. Should I add the if condition here[2] for setting dist as `ubi:${IMAGE_TAG}` for ubi image. This will make sure there is no conflict.

1. https://github.com/api7/apisix-build-tools/blob/a494c2de90ccc7973ca510b4708704555cdbbf8b/utils/determine-dist.sh#L5
2. https://github.com/api7/apisix-build-tools/blob/a494c2de90ccc7973ca510b4708704555cdbbf8b/utils/determine-dist.sh#L13

File: utils/install-common.sh (hunk 7 above)
Comment (Revolyssup) [reply utils/install-common.sh:61]: That must have been a mistake(auto formatting on my editor), I will revert this.

Comment (Revolyssup) [reply utils/install-common.sh]: Reverted

**Comment by tao12345666333** [comment]
@Revolyssup  you should add if condition. but Redhat ubi 8 should use the `el8` dist value.

**Comment by Revolyssup** [comment]
> @Revolyssup you should add if condition. but Redhat ubi 8 should use the `el8` dist value.

@tao12345666333 Okay but dist will be `el8` for centos 8 as well producing same package name. So how can I avoid the conflict here?

**Comment by tao12345666333** [comment]
In fact you just need to avoid the situation where they conflict in the cache, in the yum repo they are in different directories.

I  think you can add `ubi` to the package name.

File: utils/build-common.sh [hunk 8]
```diff
@@ -18,6 +18,10 @@ build_apisix_base_rpm() {
         set +eu
         source /opt/rh/gcc-toolset-9/enable
         set -eu
+    else
+        dnf install -y yum-utils
+        yum update --disablerepo=* --enablerepo=ubi-8-appstream-rpms --enablerepo=ubi-8-baseos-rpms -y && rm -rf /var/cache/yum
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply utils/build-common.sh]: removed it

File: utils/install-common.sh [hunk 9]
```diff
@@ -18,9 +18,15 @@ install_apisix_dependencies_rpm() {
 
 install_dependencies_rpm() {
     # install basic dependencies
-    yum -y install wget tar gcc automake autoconf libtool make curl git which unzip sudo
-    yum -y install epel-release
-    yum install -y yum-utils readline-devel
+    if [[ $IMAGE_BASE == "registry.access.redhat.com/ubi8/ubi" ]]; then
+        yum update --disablerepo=* --enablerepo=ubi-8-appstream-rpms --enablerepo=ubi-8-baseos-rpms -y && rm -rf /var/cache/yum
```
[the remark this answers is not in the document]
Comment (Revolyssup) [reply utils/install-common.sh]: removed it

[the remark this answers is not in the document]
Comment (Revolyssup) [reply utils/install-common.sh]: @tao12345666333 any other required changes?

File: .github/workflows/package-apisix-rpm-ubi.yml [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+          code=$(curl -k -i -m 20 -o /dev/null -s -w %{http_code} http://127.0.0.1:9180/apisix/admin/routes -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1')
+          if [ ! $code -eq 200 ]; then
+              echo "failed: failed to install Apache APISIX by rpm"
+              exit 1
+          fi
+
+      - name: stop and uninstall APISIX(-remote) master
+        run: |
+          docker exec ubiInstance bash -c 'PIDS=$(pgrep -f nginx); for PID in $PIDS; do kill -TERM $PID; done'
+          docker exec ubiInstance bash -c "yum -y erase apisix-remote-master"
```
Comment (tao12345666333) [inline .github/workflows/package-apisix-rpm-ubi.yml:92]: This part seems to be repeated from the above section, right?   L58～69

File: utils/install-common.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (tao12345666333) [reply utils/install-common.sh]: https://github.com/api7/apisix-build-tools/pull/296#discussion_r1198547454

**Comment by tao12345666333** [comment]
https://github.com/api7/apisix-build-tools/pull/296#discussion_r1198547454   Other than this, I have no other opinions.

File: .github/workflows/package-apisix-rpm-ubi.yml (hunk 10 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-rpm-ubi.yml:92]: @tao12345666333  You're right. And this duplication is in other workflows as well. Should I remove this duplication of 3 steps from other workflows as well?

Comment (Revolyssup) [reply .github/workflows/package-apisix-rpm-ubi.yml]: Removed duplicate steps

**Review by tao12345666333 — APPROVED** [review]
LGTM

File: .github/workflows/package-apisix-base-rpm-ubi.yml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    tags:
+      - "v*"
+    paths-ignore:
+      - '*.md'
+  pull_request:
+    branches: [ master ]
+    paths-ignore:
+      - '*.md'
+  schedule:
+    - cron: '0 0 * * *'
```
Comment (leslie-tsang) [inline .github/workflows/package-apisix-base-rpm-ubi.yml:15]: Seems we don't need to schedule build. :)

File: .github/workflows/package-apisix-rpm-ubi.yml [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    tags:
+      - "v*"
+    paths-ignore:
+      - '*.md'
+  pull_request:
+    branches: [ master ]
+    paths-ignore:
+      - '*.md'
+  schedule:
+    - cron: '0 0 * * *'
```
Comment (leslie-tsang) [inline .github/workflows/package-apisix-rpm-ubi.yml:15]: Ditto

File: utils/install-common.sh [hunk 13]
```diff
@@ -18,9 +18,14 @@ install_apisix_dependencies_rpm() {
...
-    yum -y install wget tar gcc automake autoconf libtool make curl git which unzip sudo
-    yum -y install epel-release
-    yum install -y yum-utils readline-devel
+    if [[ $IMAGE_BASE == "registry.access.redhat.com/ubi8/ubi" ]]; then
+        yum -y install --disablerepo=* --enablerepo=ubi-8-appstream-rpms --enablerepo=ubi-8-baseos-rpms wget tar gcc automake autoconf libtool make curl git which unzip sudo
+        yum install -y --disablerepo=* --enablerepo=ubi-8-appstream-rpms --enablerepo=ubi-8-baseos-rpms yum-utils
+    else 
+        yum -y install wget tar gcc automake autoconf libtool make curl git which unzip sudo
+        yum -y install epel-release
+        yum install -y yum-utils readline-devel
```
Comment (leslie-tsang) [inline utils/install-common.sh:27]: Plz use `yum install -y` to install the deps.

Comment (Revolyssup) [reply utils/install-common.sh:27]: Functionally the flag can be put anywhere. Okay I will put the flag after `install`. Thank you.

File: .github/workflows/package-apisix-base-rpm-ubi.yml (hunk 11 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-base-rpm-ubi.yml:15]: This same schedule trigger exists on other workflows as well. Should I remove them too?

File: utils/install-common.sh (hunk 13 above)
Comment (Revolyssup) [reply utils/install-common.sh]: done

File: .github/workflows/package-apisix-base-rpm-ubi.yml (hunk 11 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-base-rpm-ubi.yml]: done

File: .github/workflows/package-apisix-rpm-ubi.yml (hunk 12 above)
Comment (Revolyssup) [reply .github/workflows/package-apisix-rpm-ubi.yml]: removed

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**